src/Modules/OrchardCore.Commerce/Controllers/PaymentController.cs

```diff
@@ -245,70 +223,44 @@ public async Task<IActionResult> SuccessPost(string orderId)
             await workflowManager.TriggerEventAsync(nameof(OrderCreatedEvent), order, "Order-" + order.ContentItemId);
         }
 
-        return Redirect($"~/success/{order.ContentItemId}");
-    }
+        var currentShoppingCart = await _shoppingCartPersistence.RetrieveAsync();
+        currentShoppingCart?.Items?.Clear();
 
-    [Route(nameof(ConfirmPayment))]
-    [HttpPost]
-    [ValidateAntiForgeryToken]
-    public async Task<IActionResult> ConfirmPayment(string paymentId)
-    {
-        PaymentIntent paymentIntent;
-
-        try
-        {
-            paymentIntent = await _stripePaymentService.GetPaymentIntentAsync(paymentId);
-        }
-        catch (StripeException exception)
-        {
-            return Json(new { error = exception.StripeError.Message });
-        }
-        catch (Exception exception)
-        {
-            _logger.LogError(exception, "Payment processing failed.");
-            var message = T["An error has occurred while processing the payment. Please verify and try again."];
-            return Json(new { error = message.Value });
-        }
+        // Shopping cart ID is null by default currently.
+        await _shoppingCartPersistence.StoreAsync(currentShoppingCart);
 
-        return await GeneratePaymentResponseAsync(paymentIntent);
+        // Set back to default, because a new payment intent should be created on the next checkout.
+        _paymentIntentPersistence.Store(paymentIntentId: string.Empty);
     }
 
-    private async Task<IActionResult> GeneratePaymentResponseAsync(PaymentIntent paymentIntent)
+    [AllowAnonymous]
+    [HttpGet("checkout/middleware")]
+    // Sadly that is how it is given.
+#pragma warning disable CA1707
+    public async Task<IActionResult> PaymentConfirmationMiddleware([FromQuery] string payment_intent = null)
+#pragma warning restore CA1707
     {
-        if (paymentIntent.Status == PaymentIntentStatuses.Succeeded)
-        {
-            // The payment didn't need any additional actions and completed!
-            // Create the order content item.
-            var order = await _stripePaymentService.CreateOrderFromShoppingCartAsync(paymentIntent);
+        var fetchedPaymentIntent = await _stripePaymentService.GetPaymentIntentAsync(payment_intent);
+        var orderId = (await _stripePaymentService.GetOrderPaymentByPaymentIntentIdAsync(payment_intent))?.OrderId;
 
-            return Json(new { Success = true, OrderContentItemId = order.ContentItemId });
-        }
+        var order = await _contentManager.GetAsync(orderId);
+        var status = order?.As<OrderPart>()?.Status?.Text;
+        var finished = fetchedPaymentIntent.Status == PaymentIntentStatuses.Succeeded &&
+                       !string.IsNullOrEmpty(orderId) &&
```

Comment:
This should be quicker if you have to fetch `order` anyway.
```suggestion
                       order != null &&
```